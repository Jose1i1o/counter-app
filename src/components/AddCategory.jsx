import { useState } from 'react';
import './addCategory.css';
import PropTypes from 'prop-types';


const AddCategory = ({ setCategories }) => {

  const [ inputValue, setInputValue ] = useState('');
  
  const handleInputChange = (e) => {
    setInputValue(e.target.value);
  }

  const handleSubmit = (e) => {
    e.preventDefault();

    if(inputValue.trim().length > 1 ) {
    setCategories(categories => [...categories, inputValue ]);
    setInputValue('');
    }
    else {
      setInputValue('Try inserting a new input')
    }
  }

  return (
    <form onSubmit={ handleSubmit } >
      <input
        className="add-category__input"
        type="text"
        value={ inputValue }
        onChange={ handleInputChange }
      >
      </input>
    </form>
  )
}

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
}

export default AddCategory;