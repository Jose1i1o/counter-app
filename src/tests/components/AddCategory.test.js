// Enzyme not available for React 18. Tested with React 17

import React from 'react';
import '@testing-library/jest-dom';

import { shallow } from 'enzyme';
import AddCategory from '../../components/AddCategory';

    // const setCategories = () => {};
    const setCategories = jest.fn(); // Jest implements the above with .fn()
    let wrapper = shallow( <AddCategory setCategories={ setCategories } /> );

    // beforeEach will reset all tests before moving to the next
    beforeEach( () => {
        jest.clearAllMocks();
        wrapper = shallow( <AddCategory setCategories={ setCategories } /> );
    });

    
    test('to display correctly', () => {
        expect( wrapper ).toMatchSnapshot();
    })

    test('to change text box', () => {

        const input = wrapper.find('input');
        const value = 'Hello World';

        input.simulate('change', { target: { value } });

        expect( wrapper.find('p').text().trim() ).toBe( value );
        
    })

    test('prevent submit information on submit', () => {
        
        wrapper.find('form').simulate('submit', { preventDefault(){} }); // preventDefault(){} is the same as preventDefault = () => {}

        expect( setCategories ).not.toHaveBeenCalled();

    })
    
    test('call setCategories and clean the text box', () => {

        const value = 'Hello World';

        // 1. simulate inputChange
        wrapper.find('input').simulate('change', { target: { value } });

        // 2. simulate submit
        wrapper.find('form').simulate('submit', { preventDefault(){} });

        // 3. setCategories should be called
        expect( setCategories ).toHaveBeenCalled();
        expect( setCategories ).toHaveBeenCalledTimes(1);
        expect( setCategories ).toHaveBeenCalledWith( expect.any(Function)  ); // To check that a function has been called

        // 4. Input value should be ''
        expect( wrapper.find('input').prop('value') ).toBe('');

    })
