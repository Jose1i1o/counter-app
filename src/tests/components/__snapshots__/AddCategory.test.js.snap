// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <GifGrid /> to display correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <p>
     
     
  </p>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
